integration/spark-interfaces-scala/src/main/scala/io/openlineage/spark/builtin/scala/v1/columnLevel.scala

```diff
@@ -0,0 +1,76 @@
+package io.openlineage.spark.builtin.scala.v1
+
+import io.openlineage.client.utils.DatasetIdentifier
+
+trait ColumnLevelLineageNode {
+  def columnLevelLineageInputs(
+      context: OpenLineageContext
+  ): List[DatasetFieldLineage] = List()
+  def columnLevelLineageOutputs(
+      context: OpenLineageContext
+  ): List[DatasetFieldLineage] = List()
+  def columnLevelLineageDependencies(
+      context: OpenLineageContext
+  ): List[ExpressionDependency] = List()
+}
+
+sealed trait DatasetFieldLineage
+
+/** Output field requires only field name and expression identifier. Proper
+  * dataset identifier can be extracted from output dataset of the job available
+  * when constructing column lineage facet. This is possible as only one field
+  * with the same name is allowed in output for Spark.
+  *
+  * @param field
+  */
+case class OutputDatasetField(field: String, exprId: OlExprId)
+    extends DatasetFieldLineage
+
+sealed trait FieldWithExprId extends DatasetFieldLineage {
+  def field: String;
+  def exprId: OlExprId;
+}
+
+trait DatasetFieldWithIdentifier extends DatasetFieldLineage {
+  def datasetIdentifier: DatasetIdentifier;
+  def fieldWithExprId: FieldWithExprId
+}
+
+case class InputDatasetFieldWithIdentifier(
+    datasetIdentifier: DatasetIdentifier,
+    field: String,
+    exprId: OlExprId
+) extends DatasetFieldLineage
+    with FieldWithExprId
+
+case class InputDatasetFieldFromDelegate(
```

Comment:
Why do only `InputDatasetFields` have the identifier/delegate version? Is this because we assume `OutputDatasetFields`  to be used within some `LogicalPlan` context?

Reply:
This has to do with the way column lineage is working for spark. The entry point is method:
```
 public static Optional<OpenLineage.ColumnLineageDatasetFacet> buildColumnLineageDatasetFacet(
      OpenLineageContext context, OpenLineage.SchemaDatasetFacet schemaFacet)
```
which takes schema facet as an argument. So there is no need to derive output dataset fields as they're provided.